[Chew Ee Jia] iP

cheweejia opened the pull request:
### DukeBoi

As my mother used to say:
> Duke is love, duke is life.

DukeBoi is ~~so-so~~ excellent. It is
* quite fast
* quite easy to learn
* **singaporean**

All you need to do is,
1. Download the jar file [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ) and run in your command line
2. Try to figure out how to input the commands correctly
3. Screw up and quit 🙃

Features:
- [ ] Add stuff
- [ ] Find stuff
- [ ] Delete stuff

Here is some code to give you some PTSD
```
function functional_expressionism() {
   const decrement_repeater =
	   repeater =>
		   repeater((iter_count, x) => iter_count)(zero_repeater);
}
```

Commit:
Add Gradle support

cheweejia committed:
added ChatBot.java

cheweejia committed:
Merge remote-tracking branch 'origin/add-gradle-support'

cheweejia committed:
Added Addable.java

cheweejia committed:
Shifted the implementation of ChatBot.java to Duke.java

cheweejia committed:
Removed Addable.java

cheweejia committed:
Implemented Task class to represent task objects

cheweejia committed:
Edited Duke.java to handle Task objects

cheweejia committed:
Implemented subclasses of Task

cheweejia committed:
Implemented support for subclasses of tasks

cheweejia committed:
Added input and the expected output for the Automated Text UI Testing

cheweejia committed:
Bug fixes

cheweejia committed:
Implemented DukeException

cheweejia committed:
Support handling of DukeExceptions

cheweejia committed:
Added support for deleting tasks from list

cheweejia committed:
Implement save feature

cheweejia committed:
Add date and time management

cheweejia committed:
Merge branch 'branch-Level-8'

cheweejia committed:
Implement more OOP

cheweejia committed:
Package the files

cheweejia committed:
Implement JUnit tests

cheweejia committed:
Add JavaDoc

cheweejia committed:
Improve standard of code

cheweejia committed:
Implement find feature

cheweejia committed:
Merge branch 'branch-A-CodingStandard'

cheweejia committed:
Merge branch 'branch-Level-9'

cheweejia committed:
Update README.md

cheweejia committed:
Update README.md

cheweejia committed:
Update README.md

cheweejia committed:
Add Gradle support

cheweejia committed:
Merge remote-tracking branch 'origin/master'

cheweejia committed:
Implement CheckStyle

yewmeng28 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,50 @@
+     * Marks the task as completed.
+     */
+    public void markAsDone() {
+        isDone = !isDone;
```
Perhaps it would be better to simply change `isDone` to `true`, in the case that the code accidentally triggers the `markAsDone()` again, causing the `isDone` to become false again.

yewmeng28 reviewed (commented):
Only one very minor comment in your code. Looks good!

cheweejia committed:
Implement GUI

rldefa commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,36 @@
+
+    /**
+     * Constructs a Deadline object.
+     * @param description Description of the deadline.
```
Should there be an empty line between description and param?

rldefa commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,36 @@
+ * Implements an Event object extending from Task.
+ */
+public class Event extends Task {
+    protected final LocalDateTime dateTime;
```
Should the variable name be more specific? like eventDateTime.

rldefa commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+        return list;
+    }
+
+    private void loadToDo(String data, ArrayList<Task> list) {
```
Should the parameter name be more specific? What is data and what is list.

yyhangz commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,44 @@
+                } else if (input.contains("delete")) {
+                    return tasks.handleDelete(input, storage, ui);
+                } else if (input.contains("find")) {
+                    return tasks.handleFind(input, ui);
```
Perhaps it may be good to consider an alternative to `String::contains` in order to account for edge cases such as `event finalize project deadline /at *time*`, since this will be handled as a `deadline` command instead of an `event` command.

yyhangz commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,164 @@
+     */
+    public String handleDone(String input, Storage storage, Ui ui) throws DukeException {
+        try {
+            int taskNumber = Integer.parseInt(input.substring(5));
```
It might be worth to consider defining the commands' string length as class level constants to avoid the pitfall of magic literals, as well as for easier refactoring, should your command change from `done` to a command of a different string length in the future.

yyhangz commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,164 @@
+     */
+    public String handleDeadline(String input, Storage storage, Ui ui) throws DukeException {
+        try {
+            int split = input.indexOf("/");
```
Perhaps there could be a more specific name for the variable `split`, like `splitAtIdx`?

yyhangz commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,132 @@
+     * @param task The deleted Task.
+     * @param length The number of Tasks in the TaskList.
+     */
+    public String delete(Task task, int length) {
```
Similarly, perhaps the name for this method could be more specific?

yyhangz reviewed (commented):
Looks good overall!1!1

cheweejia committed:
Implement assertions

The program does not utilise assertion.

Assertions can help to identify bugs and improve program, by documenting important assumptions that should hold at various points in the code.

Let's introduce assertions into our program to identify the state of the file before performing any actions on it.

Using assertions in addition to JUnit tests can help to ensure that the program is functional during runtime.

cheweejia committed:
Merge pull request #1 from cheweejia/branch-A-Assertions

Implement assertions

cheweejia committed:
Improve code quality

Certain parts of the program contain deep nesting and magic numbers.

Improving the code quality makes the program more readable and easy to understand.

Let's remove all the magic numbers, and redo the logic of some of the implements. Let's also abide by SLAP and make the happy path prominent.

Following the code quality standards strictly rather than going by intuition can guarantee a better product.

cheweejia committed:
Merge pull request #2 from cheweejia/branch-A-CodeQuality

Improve code quality

cheweejia committed:
Create android.yml

cheweejia committed:
Implement tagging of tasks

Tasks can only be searched with keywords associated with its description.

Search function is not very intuitive and useful, since users have to remember the tasks by its description in order to find it in the future.

Let's introduce tagging as a way to classify tasks.

Tagging helps to improve the search function, since users can now search for the category of tasks, rather than a keyword from the description. This will improve usability.

cheweejia committed:
Resolve gradle build error with regards to image

cheweejia committed:
Merge pull request #3 from cheweejia/branch-C-Tagging

Implement tagging of tasks

cheweejia committed:
Edit styling

cheweejia committed:
Update User Guide

cheweejia committed:
Fix JAR

cheweejia committed:
Merge branch 'master' of https://github.com/cheweejia/ip
